canisters/router/src/upgrade.rs

```diff
@@ -1,11 +1,78 @@
-use crate::Data;
-use ic_cdk::api::stable::StableWriter;
+use crate::{get_user_root_buckets, Data};
+use ic_cdk::api::stable::{StableReader, StableWriter};
+use ic_history_common::{GetUserRootBucketsArg, RootBucketId};
+use ic_kit::candid::{candid_method, encode_args};
 use ic_kit::ic;
-use ic_kit::macros::pre_upgrade;
+use ic_kit::macros::{post_upgrade, pre_upgrade, update};
+use ic_kit::Principal;
+
+#[derive(Default)]
+struct RootBucketsToUpgrade(Vec<RootBucketId>);
 
 #[pre_upgrade]
 fn pre_upgrade() {
     let data = ic::get::<Data>();
     let writer = StableWriter::default();
     serde_cbor::to_writer(writer, &data).expect("Failed to serialize data.");
 }
+
+#[post_upgrade]
+fn post_upgrade() {
+    let reader = StableReader::default();
+    let data = serde_cbor::from_reader(reader).expect("Failed to deserialize");
+    ic::store::<Data>(data);
+
+    let root_buckets = get_user_root_buckets(GetUserRootBucketsArg {
+        user: Principal::management_canister(),
+        witness: false,
+    })
+    .contracts;
+
+    ic::store(RootBucketsToUpgrade(root_buckets.to_vec()));
+}
+
+#[update]
+#[candid_method(update)]
+fn trigger_upgrade() {
+    let canisters = ic::get_mut::<RootBucketsToUpgrade>();
+
+    if canisters.0.is_empty() {
+        return;
+    }
+
+    for _ in 0..16 {
```

Comment:
Is there a specific reason to do it in batches of 16?

Reply:
follow

Reply:
no particular reason for it to be 16.  it just seemed like a reasonable number as it's not that small to take many level of nesting for completion and it's not too large to fail for a random reason.